orangecontrib/text/widgets/owstatistics.py

```diff
@@ -0,0 +1,685 @@
+    AlmostPerfectContextHandler,
+)
+
+
+def num_words(document: str, callback: Callable) -> int:
+    """
+    Return number of words in document-string. Word is every entity divided by
+    space, tab, newline.
+    """
+    callback()
+    return len(document.split())
+
+
+def char_count(document: str, callback: Callable) -> int:
+    """
+    Count number of alpha-numerical in document/string.
+    """
+    callback()
+    return sum(c.isalnum() for c in document)
+
+
+def digit_count(document: str, callback: Callable) -> int:
+    """
+    Count number of digits in document/string.
+    """
+    callback()
+    return sum(c.isdigit() for c in document)
+
+
+def count_appearances(
+    document: str, characters: List[str], callback: Callable
+) -> int:
+    """
+    Count number of appearances of chars from `characters` list.
+    """
+    callback()
+    # I think it supports the majority of main languages
+    # Y can be vowel too sometimes - it is not possible to distinguish
+    return sum(document.lower().count(c) for c in characters)
+
+
+def preprocess_only_words(corpus: Corpus) -> Corpus:
+    """
+    Apply the preprocessor that splits words, transforms them to lower case
+    (and removes punctuations).
+
+    Parameters
+    ----------
+    corpus
+        Corpus on which the preprocessor will be applied.
+
+    Returns
+    -------
+    Preprocessed corpus. Result of pre-processing is saved in tokens/ngrams.
+    """
+    p = Preprocessor(
+        transformers=[LowercaseTransformer()],
+        # by default regexp keeps only words (no punctuations, no spaces)
+        tokenizer=RegexpTokenizer(),
+    )
+    return p(corpus, inplace=False)
+
+
+# every statistic returns a np.ndarray with statistics
+# and list with variables names - it must be implemented here since some
+# statistics in the future will have more variables
+
+
+def words_count(
+    corpus: Corpus, _: str, callback: Callable
+) -> Tuple[np.ndarray, List[str]]:
+    """
+    Count number of words in each document.
+    """
+    corpus = preprocess_only_words(corpus)
+    # np.c_ makes column vector (ndarray) out of the list
+    # [1, 2, 3] -> [[1], [2], [3]]
+    return (
+        np.c_[[num_words(d, callback) for d in corpus.documents]],
+        ["Word count"],
+    )
+
+
+def characters_count(
+    corpus: Corpus, _: str, callback: Callable
+) -> Tuple[np.ndarray, List[str]]:
+    """
+    Count number of characters without spaces, newlines, tabs, ...
+    """
+    return (
+        np.c_[[char_count(d, callback) for d in corpus.documents]],
+        ["Character count"],
+    )
+
+
+def n_gram_count(
+    corpus: Corpus, _: str, callback: Callable
+) -> Tuple[np.ndarray, List[str]]:
+    """
+    Count number of n-grams in every document
+    """
+
+    def ng_count(n_gram: List[str]):
+        callback()
+        return len(n_gram)
+
+    return np.c_[list(map(ng_count, corpus.ngrams))], ["N-gram count"]
+
+
+def word_density(
```

Comment:
It would be nice to update function call to reflect the function --> average_word_len or something similar. Also please change the docstring.